### PR TITLE
Create new empty span stack in ScopeManager.Activate() if threadLocalSpanStack is null

### PR DESCRIPTION
**Pull request by fstien**
Removes a bug preventing the activation of spans when there are none currently active.

### CODE REVIEW

File: src/main/kotlin/com/zopa/ktor/opentracing/Span.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    val tracer = getGlobalTracer()
+    val span = tracer.buildSpan(name).start()
+    try {
+        tracer.scopeManager().activate(span).use { scope ->
+            return block(span)
+        }
+    } finally {
+        span.finish()
+    }
+}
```
Comment (fstien) [inline src/main/kotlin/com/zopa/ktor/opentracing/Span.kt:15]: Moving this out of Utils.kt and into it's own file

File: src/main/kotlin/com/zopa/ktor/opentracing/Coroutines.kt [hunk 2]
```diff
@@ -14,11 +14,13 @@ import kotlin.coroutines.EmptyCoroutineContext
 
 
 fun tracingContext(): CoroutineContext {
-    val activeSpan: Span? = threadLocalSpanStack.get()?.peek()
+    val activeSpan: Span? = getGlobalTracer().scopeManager().activeSpan()
+
```
Comment (fstien) [inline src/main/kotlin/com/zopa/ktor/opentracing/Coroutines.kt:18]: Re-use the read from the thread local in the Scope manager implementation here; https://github.com/zopaUK/ktor-opentracing/blob/7709211b5716bd50801d710c517258f3f00c452c/src/main/kotlin/com/zopa/ktor/opentracing/ThreadContextElementScopeManager.kt#L23

File: src/main/kotlin/com/zopa/ktor/opentracing/ThreadContextElementScopeManager.kt [hunk 3]
```diff
@@ -3,26 +3,27 @@ package com.zopa.ktor.opentracing
...
 class ThreadContextElementScopeManager: ScopeManager {
     override fun activate(span: Span?): Scope {
-        val spanStack = threadLocalSpanStack.get()
+        var spanStack = threadLocalSpanStack.get()
+
         if (spanStack == null) {
-            log.error { "spanStack is null" }
-            return NoopScopeManager.NoopScope.INSTANCE
+            spanStack = Stack<Span>()
+            threadLocalSpanStack.set(spanStack)
```
Comment (fstien) [inline src/main/kotlin/com/zopa/ktor/opentracing/ThreadContextElementScopeManager.kt:17]: Currently, if you try to activate a span without having any existing spans ( ie a null threadLocalSpanStack ) we return a Noop scope and do not store the span in the span stack.

File: src/main/kotlin/com/zopa/ktor/opentracing/ThreadContextElementScopeManager.kt [hunk 4]
```diff
@@ -34,7 +35,9 @@ internal class CoroutineThreadLocalScope: Scope {
             return
         }
 
-        spanStack.pop()
+        if (spanStack.isNotEmpty()) {
+            spanStack.pop()
+        }
```
Comment (fstien) [inline src/main/kotlin/com/zopa/ktor/opentracing/ThreadContextElementScopeManager.kt:40]: Eliminates the possibility of having a `EmptyStackException` for whatever reason

File: src/main/kotlin/com/zopa/ktor/opentracing/ThreadContextElementScopeManager.kt (hunk 3 above)
Comment (hj1611) [reply src/main/kotlin/com/zopa/ktor/opentracing/ThreadContextElementScopeManager.kt:17]: Might be worth adding a quick unit test to cover this functionality to ensure that the bug is indeed fixed?

Also could add an info log to indicate when this code path is being run through.

Comment (fstien) [reply src/main/kotlin/com/zopa/ktor/opentracing/ThreadContextElementScopeManager.kt:18]: Yes good idea , added a test here https://github.com/zopaUK/ktor-opentracing/pull/17/commits/a74c16a45793663171213cb5c48042323fc8e2f9 and the info log here https://github.com/zopaUK/ktor-opentracing/pull/17/commits/b808d224f381dd7742db3f8f059099bb0b95ac4a

**Review by hj1611 — APPROVED** [review]

**Pull request merged**